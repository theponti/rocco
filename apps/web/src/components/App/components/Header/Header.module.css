.header {
  @apply bg-blue-300;
  display: flex;
  align-content: center;
  justify-content: space-between;
  margin: 0 0 40px;
  padding: 24px;
  /* Blur background */
  backdrop-filter: blur(12px);
  background-color: rgba(247, 247, 247, 0.8);
}

.appName {
  font-weight: 700;
}

.link {
  text-decoration: none;
}
