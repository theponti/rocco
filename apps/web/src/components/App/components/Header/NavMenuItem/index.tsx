import styles from "./NavMenuItem.module.css";
import React from "react";

function NavMenuItem({
  children,
  ...props
}: React.HTMLAttributes<HTMLDivElement>) {
  return (
    <div className={styles.menuItem} {...props}>
      {children}
    </div>
  );
}

export default NavMenuItem;
