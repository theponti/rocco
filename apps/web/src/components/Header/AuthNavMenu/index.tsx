import { List, LogOut, Mail, Search, Settings, UserCircle } from "lucide-react";
import { useCallback, useEffect, useRef } from "react";
import { Link, useNavigate } from "react-router-dom";

import AppLink from "src/components/AppLink";
import { useAuth } from "src/lib/auth";
import type { User } from "src/lib/types";
import {
	ACCOUNT,
	DASHBOARD,
	INVITES,
	LANDING,
	LISTS,
} from "src/lib/utils/routes";

const AuthNavMenu = ({ user }: { user: User }) => {
	const { logout } = useAuth();
	const navMenuRef = useRef(null);
	const navigate = useNavigate();
	const onLogoutClick = useCallback(
		async () => logout.mutateAsync(),
		[logout.mutateAsync],
	);

	useEffect(() => {
		logout.status === "success" && navigate(LANDING);
	}, [logout.status, navigate]);

	const onLinkClick = useCallback(() => {
		navMenuRef.current?.click?.();
	}, []);

	const onLinkKeyDown = useCallback((event) => {
		if (event.key === "Enter") {
			event.preventDefault();
			navMenuRef.current?.click?.();
		}
	}, []);

	return (
		<>
			<div className="hidden md:visible md:flex gap-6 mr-4">
				<AppLink to={DASHBOARD} className="px-4 py-2 rounded-xl">
					<Search />
				</AppLink>
				<AppLink to={LISTS} className="px-4 py-2 rounded-xl">
					<List />
				</AppLink>
				<AppLink to={INVITES} className="px-4 py-2 rounded-xl">
					<Mail />
				</AppLink>
			</div>
			<div>
				<details data-testid="dropdown" className="dropdown dropdown-end">
					<summary
						ref={navMenuRef}
						data-testid="dropdown-button"
						role="button"
						className="btn bg-transparent border-black"
					>
						<UserCircle />
					</summary>
					<ul
						className="p-1 pb-3 shadow menu dropdown-content z-[1] bg-base-100 rounded-box w-52 mt-1"
						onKeyDown={onLinkKeyDown}
						onClick={onLinkClick}
					>
						<li>
							<span className="flex justify-end text-sm text-secondary hover:outline-none">
								{user.email}
							</span>
						</li>
						<li>
							<Link to={LISTS}>
								<List size={20} />
								Lists
							</Link>
						</li>
						<li>
							<Link to={INVITES}>
								<Mail size={20} />
								Invites
							</Link>
						</li>
						<li>
							<Link to={ACCOUNT}>
								<Settings size={20} />
								Account
							</Link>
						</li>
						<li>
							<button
								type="button"
								className="btn-ghost"
								onClick={onLogoutClick}
							>
								<LogOut size={20} />
								Logout
							</button>
						</li>
					</ul>
				</details>
			</div>
		</>
	);
};

export default AuthNavMenu;
