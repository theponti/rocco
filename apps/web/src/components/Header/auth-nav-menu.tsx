import { List, LogOut, Mail, Search, Settings, UserCircle } from "lucide-react";
import { useCallback, useEffect, useRef } from "react";
import { Link, useNavigate } from "react-router-dom";

import AppLink from "src/components/AppLink";
import {
	DropdownMenu,
	DropdownMenuContent,
	DropdownMenuItem,
	DropdownMenuTrigger,
} from "src/components/ui/dropdown-menu";
import { useAuth } from "src/lib/auth";
import { ACCOUNT, DASHBOARD, INVITES, LANDING, LISTS } from "src/lib/routes";
import type { User } from "src/lib/types";

const AuthNavMenu = ({ user }: { user: User }) => {
	const { logout } = useAuth();
	const navMenuRef = useRef(null);
	const navigate = useNavigate();
	const onLogoutClick = useCallback(
		async () => logout.mutateAsync(),
		[logout.mutateAsync],
	);

	useEffect(() => {
		logout.status === "success" && navigate(LANDING);
	}, [logout.status, navigate]);

	return (
		<>
			<div className="hidden md:visible md:flex gap-6 mr-4">
				<AppLink to={DASHBOARD} className="px-4 py-2 rounded-xl">
					<Search />
				</AppLink>
				<AppLink to={LISTS} className="px-4 py-2 rounded-xl">
					<List />
				</AppLink>
				<AppLink to={INVITES} className="px-4 py-2 rounded-xl">
					<Mail />
				</AppLink>
			</div>
			<div>
				<DropdownMenu>
					<DropdownMenuTrigger
						ref={navMenuRef}
						data-testid="auth-dropdown-button"
						role="button"
						className="btn bg-transparent border-black py-2 px-3"
					>
						<UserCircle />
					</DropdownMenuTrigger>
					<DropdownMenuContent
						data-testid="dropdown-content"
						className="min-w-[200px] w-fit"
					>
						<div className="text-right bg-white text-sm text-gray-400 hover:outline-none pt-2 px-2 pb-3 w-full">
							{user.email}
						</div>
						<DropdownMenuItem>
							<Link to={LISTS} className="flex gap-4 py-2">
								<List size={20} />
								Lists
							</Link>
						</DropdownMenuItem>
						<DropdownMenuItem>
							<Link to={INVITES} className="flex gap-4 py-2">
								<Mail size={20} />
								Invites
							</Link>
						</DropdownMenuItem>
						<DropdownMenuItem>
							<Link to={ACCOUNT} className="flex gap-4 py-2">
								<Settings size={20} />
								Account
							</Link>
						</DropdownMenuItem>
						<DropdownMenuItem>
							<button
								type="button"
								className="btn-ghost flex gap-4 py-2"
								onClick={onLogoutClick}
							>
								<LogOut size={20} />
								Logout
							</button>
						</DropdownMenuItem>
					</DropdownMenuContent>
				</DropdownMenu>
			</div>
		</>
	);
};

export default AuthNavMenu;
