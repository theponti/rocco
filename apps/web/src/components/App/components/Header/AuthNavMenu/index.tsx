import styled from "@emotion/styled";
import * as DropdownMenu from "@radix-ui/react-dropdown-menu";
import { HamburgerMenuIcon } from "@radix-ui/react-icons";
import { useCallback } from "react";
import { useNavigate } from "react-router-dom";

import { ACCOUNT_PATH, LANDING_PATH } from "src/constants/routes";
import { logout } from "src/services/auth";
import { User } from "src/services/auth/auth.types";
import Avatar from "ui/Avatar";
import NavMenuItem from "ui/NavMenuItem";

import "./styles.css";
import { useAppDispatch } from "src/services/hooks";

const Wrap = styled(NavMenuItem)`
  display: flex;
  gap: 10px;
  border: 1px solid slateblue;
  border-radius: 50px;
  padding: 4px;
  margin: -4px -4px 0 0;

  &:hover {
    box-shadow: 0px 2px 5px 0px slateblue;
  }
`;

const AuthNavMenu = ({ user }: { user: User }) => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const onAccountClick = useCallback(() => {
    navigate(ACCOUNT_PATH);
  }, [navigate]);
  const onLogoutClick = useCallback(() => {
    dispatch(logout()).then(() => {
      navigate(LANDING_PATH);
    });
  }, []);

  return (
    <Wrap>
      <Avatar />
      <DropdownMenu.Root>
        <DropdownMenu.Trigger asChild>
          <button className="IconButton" aria-label="account menu">
            <HamburgerMenuIcon />
          </button>
        </DropdownMenu.Trigger>
        <DropdownMenu.Portal>
          <DropdownMenu.Content
            className="DropdownMenuContent"
            align="end"
            sideOffset={5}
          >
            <DropdownMenu.Item
              className="DropdownMenuItem"
              onClick={onAccountClick}
            >
              Account
            </DropdownMenu.Item>
            <DropdownMenu.Separator className="DropdownMenuSeparator" />
            <DropdownMenu.Item
              className="DropdownMenuItem"
              onClick={onLogoutClick}
            >
              Logout
            </DropdownMenu.Item>
          </DropdownMenu.Content>
        </DropdownMenu.Portal>
      </DropdownMenu.Root>
    </Wrap>
  );
};

export default AuthNavMenu;
