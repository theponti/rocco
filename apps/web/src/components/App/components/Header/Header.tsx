import { Link } from "react-router-dom";

import { APP_NAME } from "src/constants";
import { LANDING_PATH, LOGIN_PATH } from "src/constants/routes";

import styles from "./Header.module.css";
import AuthNavMenu from "./AuthNavMenu";
import NavMenuItem from "./NavMenuItem";
import { useAppSelector } from "src/services/hooks";
import { getIsAuthenticated } from "src/services/auth";

function Header() {
  const isAuthenticated = useAppSelector(getIsAuthenticated);
  const btnClassName = "btn border-none min-h-0 h-10 text-white ml-4";
  return (
    <header className="flex items-center justify-between py-2 px-4">
      <NavMenuItem>
        <Link to={LANDING_PATH} className={styles.appName}>
          {APP_NAME}
        </Link>
      </NavMenuItem>
      {isAuthenticated ? (
        <AuthNavMenu />
      ) : (
        <NavMenuItem>
          <Link to={LOGIN_PATH} className={btnClassName}>
            Log In
          </Link>
        </NavMenuItem>
      )}
    </header>
  );
}

export default Header;
