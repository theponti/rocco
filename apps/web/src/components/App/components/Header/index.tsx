import { Globe } from "lucide-react";
import { Link } from "react-router-dom";

import { useAuth } from "src/lib/hooks";
import { APP_NAME } from "src/lib/utils";
import { LANDING, LOGIN } from "src/lib/utils/routes";

import AuthNavMenu from "./AuthNavMenu";

function Header() {
	const { isAuthenticated, user } = useAuth();

	return (
		<header className="navbar flex flex-row justify-between items-center pt-4 px-0">
			<Link
				to={LANDING}
				className="flex gap-2 items-center font-extrabold lowercase text-4xl py-1"
			>
				<Globe className="animate-spin-slow mt-1 size-7 text-primary" />
				{APP_NAME}
			</Link>
			{isAuthenticated ? (
				<AuthNavMenu user={user} />
			) : (
				<Link to={LOGIN} className="btn btn-primary text-white">
					Log In
				</Link>
			)}
		</header>
	);
}

export default Header;
