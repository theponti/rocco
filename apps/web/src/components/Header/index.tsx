import { Globe } from "lucide-react";
import { Link } from "react-router-dom";

import { useAuth } from "src/lib/auth";
import { AuthStatus } from "src/lib/auth/types";
import { APP_NAME } from "src/lib/constants";
import { LANDING, LOGIN } from "src/lib/routes";

import AuthNavMenu from "./auth-nav-menu";

function Header() {
	const { user, status } = useAuth();

	return (
		<header className="navbar flex flex-row justify-between items-center pt-4 px-0">
			<Link
				to={LANDING}
				className="flex gap-2 items-center font-extrabold lowercase text-4xl py-1"
			>
				<Globe className="animate-spin-slow mt-1 size-7 text-primary" />
				{APP_NAME}
			</Link>
			{status === AuthStatus.Authenticated ? (
				<AuthNavMenu user={user} />
			) : (
				<Link to={LOGIN} className="btn btn-primary text-white">
					Log In
				</Link>
			)}
		</header>
	);
}

export default Header;
