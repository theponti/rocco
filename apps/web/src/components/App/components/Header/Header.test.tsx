import Header from "./Header";
import { Provider } from "react-redux";
import { MemoryRouter as Router } from "react-router-dom";
import { store } from "../../../../services/store";
import renderer from "react-test-renderer";
import { test, expect, describe } from "vitest";

describe("<Header/>", () => {
  test("should render component", () => {
    const div = renderer.create(
      <Router>
        <Provider store={store}>
          <Header />
        </Provider>
      </Router>
    );
    expect(div.toJSON()).toMatchSnapshot();
  });
});
