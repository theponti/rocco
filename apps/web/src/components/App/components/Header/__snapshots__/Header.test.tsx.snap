// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header/> > should render component 1`] = `
<header
  className="flex items-center justify-between py-2 max-md:px-2"
>
  <div
    className="_menuItem_02ca72"
  >
    <a
      className="text-primary font-semibold"
      href="/"
      onClick={[Function]}
    >
      Rocco
    </a>
  </div>
  <div
    className="_menuItem_02ca72"
  >
    <a
      className="btn bg-blue-600 border-none min-h-0 h-10 text-white ml-4"
      href="/login"
      onClick={[Function]}
    >
      Log In
    </a>
  </div>
</header>
`;
